src/bucket/Bucket.cpp

```diff
@@ -29,10 +29,274 @@
 #include <fmt/format.h>
 #include <future>
 
+#include "util/XDRCereal.h"
+
 namespace stellar
 {
 
-Bucket::Bucket(std::string const& filename, Hash const& hash)
+// Returns next page boundary > pos
+static inline size_t
+nextPageBoundary(size_t pos, size_t pageSize)
+{
+    return (pos & ~(pageSize - 1)) + pageSize;
+}
+
+// Returns a liquidity pool key, with the liquidityPoolID filled with the given
+// byte
+static LedgerKey
+getDummyTrustlineKey(uint8_t fill)
+{
+    LedgerKey key(LIQUIDITY_POOL);
+    key.liquidityPool().liquidityPoolID.fill(fill);
+    return key;
+}
+
+LedgerKey
+Bucket::getBucketLedgerKey(BucketEntry const& be)
+{
+    switch (be.type())
+    {
+    case LIVEENTRY:
+    case INITENTRY:
+        return LedgerEntryKey(be.liveEntry());
+    case DEADENTRY:
+        return be.deadEntry();
+    case METAENTRY:
+    default:
+        throw std::invalid_argument("Tried to get key for METAENTRY");
+    }
+}
+
+BucketRangeIndex::BucketRangeIndex(std::filesystem::path const& filename,
+                                   size_t pageSize)
+    : mPageSize(pageSize)
+{
+    ZoneScoped;
+    releaseAssert(!filename.empty());
+    releaseAssert(pageSize > 256);
```

Comment:
the assert on `pageSize` is probably not what you want:
* I think you intend to require to have at least 1 object per page but things will work anyways, so it's more of a tuning thing. I would say a warning is probably enough (in config.cpp) and the number should be picked so that we expect "reasonable performance" (so even 256 in that context may not be great).
* Jump Cannon introduces much larger key sizes anyways

@graydon  - as I write this I am not sure we should allow "any key size" in Jump Cannon (https://github.com/stellar/stellar-protocol/blob/master/core/cap-0053.md) , this has pretty big implications on the performance of any "in memory index". So it would be pretty bad if people craft ledger entries with 1KB keys type of thing. I think we should just force people to compute their own "short keys", we can keep `SCVal` as part of their internal key, but limit the overall key size to something reasonable like 128 bytes (contractID + 2 full keys + extra stuff seems plenty).